c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,378 @@
+    ///
+    /// This is used immediately before freeing,
+    /// before both [`Callee::Free`] and [`Callee::Realloc`].
+    To,
+}
+
+impl CVoidCastDirection {
+    /// For applicable [`Callee`]s that take or return [`*c_void`](core::ffi::c_void),
+    /// return the [`CVoidCastDirection`]s that they use for casts.
+    ///
+    /// That is, these [`Callee`]s are [`CVoidCastDirection::From`]:
+    /// * [`Callee::Malloc`]
+    /// * [`Callee::Calloc`]
+    /// * [`Callee::Realloc`]
+    ///
+    /// And these [`Callee`]s are [`CVoidCastDirection::To`]:
+    /// * [`Callee::Free`]
+    /// * [`Callee::Realloc`]
+    pub fn from_callee(callee: Option<Callee>) -> &'static [Self] {
+        let callee = match callee {
+            None => return &[],
+            Some(it) => it,
+        };
+        use CVoidCastDirection::*;
+        use Callee::*;
+        match callee {
+            Malloc | Calloc => &[From][..],
+            Realloc => &[To, From][..],
+            Free => &[To][..],
+            _ => &[],
+        }
+    }
+}
+
+/// The [`Place`] of a [`*c_void`].
+///
+/// It is checked to be a [`*c_void`] upon construction.
+///
+/// [`*c_void`]: core::ffi::c_void
+#[derive(PartialEq, Eq, Hash, Clone, Copy)]
+pub struct CVoidPtr<'tcx> {
+    place: Place<'tcx>,
+}
+
+impl<'tcx> CVoidPtr<'tcx> {
+    /// Check if a [`Place`] is really a [`*c_void`](core::ffi::c_void)
+    /// by checking its [`TyCtxt::def_path`] and [`TyCtxt::item_name`].
+    ///
+    /// This panics on failure.
+    pub fn checked(place: Place<'tcx>, local_decls: &LocalDecls<'tcx>, tcx: TyCtxt<'tcx>) -> Self {
+        let deref_ty = place
+            .ty(local_decls, tcx)
+            .ty
+            .builtin_deref(true)
+            .unwrap()
+            .ty
+            .kind();
+        assert_matches!(deref_ty, TyKind::Adt(adt, _) => {
+            assert_eq!(tcx.def_path(adt.did()).data[0].to_string(), "ffi");
+            assert_eq!(tcx.item_name(adt.did()).as_str(), "c_void");
+        });
+        Self { place }
+    }
+
+    /// Check another [`Place`] for being a [`*c_void`](core::ffi::c_void)
+    /// by comparing it to ourself, which is already checked.
+    ///
+    /// If the [`Place`] is the same [`CVoidPtr`], return it as [`Ok`];
+    /// otherwise, return the same [`Place`] as [`Err`].
+    pub fn checked_by_eq(&self, place: Place<'tcx>) -> Result<Self, Place<'tcx>> {
+        if self.place == place {
+            Ok(Self { place })
+        } else {
+            Err(place)
+        }
+    }
+}
+
+fn get_cast_place<'tcx>(rv: &Rvalue<'tcx>) -> Option<Place<'tcx>> {
+    match rv {
+        Rvalue::Cast(_, op, _) => op.place(),
+        _ => None,
+    }
+}
+
+fn get_assign_sides<'tcx, 'a>(
+    stmt: &'a Statement<'tcx>,
+) -> Option<(Place<'tcx>, &'a Rvalue<'tcx>)> {
+    let (pl, ref rv) = match &stmt.kind {
+        StatementKind::Assign(it) => Some(&**it),
+        _ => None,
+    }?;
+    Some((*pl, rv))
+}
+
+impl<'tcx> CVoidPtr<'tcx> {
```

Comment:
why have a separate `impl` block for these methods?

Reply:
It was just so the order of those helper methods were close to their use site.  But I'll move them to `util` so it won't be a problem anymore.

Reply:
Fixed in 707f6c020d381f1e017a548725138571a6f4afae.